src/node/historical_queries.h

```diff
@@ -0,0 +1,341 @@
+        if (ib.second)
+        {
+          fetch_entry_at(idx);
+        }
+      }
+
+      // Too many outstanding fetches, this one is silently ignored
+      // TODO
+    }
+
+    void fetch_entry_at(consensus::Index idx)
+    {
+      const auto it =
+        std::find(pending_fetches.begin(), pending_fetches.end(), idx);
+      if (it != pending_fetches.end())
+      {
+        // Already fetching this index
+        return;
+      }
+
+      RINGBUFFER_WRITE_MESSAGE(
+        consensus::ledger_get,
+        to_host,
+        idx,
+        consensus::LedgerRequestPurpose::HistoricalQuery);
+      pending_fetches.insert(idx);
+    }
+
+    std::optional<ccf::Signature> get_signature(const StorePtr& sig_store)
+    {
+      kv::Tx tx;
+      auto sig_table = sig_store->get<ccf::Signatures>(ccf::Tables::SIGNATURES);
+      if (sig_table == nullptr)
+      {
+        throw std::logic_error(
+          "Missing signatures table in signature transaction");
+      }
+
+      auto sig_view = tx.get_view(*sig_table);
+      return sig_view->get(0);
+    }
+
+    std::optional<ccf::NodeInfo> get_node_info(ccf::NodeId node_id)
+    {
+      // Current solution: Use current state of Nodes table from real store.
+      // This only works while entries are never deleted from this table, and
+      // makes no check that the signing node was active at the point it
+      // produced this signature
+      kv::Tx tx;
+
+      auto nodes_table = source_store.get<ccf::Nodes>(ccf::Tables::NODES);
+      if (nodes_table == nullptr)
+      {
+        throw std::logic_error("Missing nodes table");
+      }
+
+      auto nodes_view = tx.get_view(*nodes_table);
+      return nodes_view->get(node_id);
+    }
+
+    void handle_signature_transaction(
+      consensus::Index sig_idx, const StorePtr& sig_store)
+    {
+      const auto sig = get_signature(sig_store);
+      if (!sig.has_value())
+      {
+        throw std::logic_error(
+          "Missing signature value in signature transaction");
+      }
+
+      // Build tree from signature
+      ccf::MerkleTreeHistory tree(sig->tree);
+      const auto real_root = tree.get_root();
+      if (real_root != sig->root)
+      {
+        throw std::logic_error("Invalid signature: invalid root");
+      }
+
+      // TODO: Check the signature is produced by the claimed node (needs
+      // retrieval of the node's cert)
+      const auto node_info = get_node_info(sig->node);
+      if (!node_info.has_value())
+      {
+        throw std::logic_error(fmt::format(
+          "Signature {} claims it was produced by node {}: This node is "
+          "unknown",
+          sig_idx,
+          sig->node));
+      }
+
+      auto verifier = tls::make_verifier(node_info->cert);
+      const auto verified = verifier->verify_hash(
+        real_root.h.data(),
+        real_root.h.size(),
+        sig->sig.data(),
+        sig->sig.size());
+      if (!verified)
+      {
+        throw std::logic_error(
```

Comment:
Do we want to crash the service if there is an issue with the data that we read from outside the enclave?